packages/adblocker/src/engine/engine.ts

```diff
@@ -86,16 +86,29 @@ export interface Caching {
   write: (path: string, buffer: Uint8Array) => Promise<void>;
 }
 
-export default class FilterEngine extends EventEmitter<
-  | 'csp-injected'
-  | 'html-filtered'
-  | 'request-allowed'
-  | 'request-blocked'
-  | 'request-redirected'
-  | 'request-whitelisted'
-  | 'script-injected'
-  | 'style-injected'
-> {
+export type EngineEventContext<T> = {
+  hostname: string;
+  classes: string[] | undefined;
+  hrefs: string[] | undefined;
+  ids: string[] | undefined;
+  context?: T;
+};
+
+export type EngineEventHandlers = {
+  'request-allowed': (request: Request, result: BlockingResponse) => any;
+  'request-blocked': (request: Request, result: BlockingResponse) => any;
+  'request-redirected': (request: Request, result: BlockingResponse) => any;
+  'request-whitelisted': (request: Request, result: BlockingResponse) => any;
+  'html-filtered': (htmlSelectors: HTMLSelector[], url: string) => any;
+  'csp-injected': (csps: string, request: Request) => any;
+  'script-injected': <T>(script: string, url: string, context: EngineEventContext<T>) => any;
+  'style-injected': <T>(script: string, url: string, context: EngineEventContext<T>) => any;
+  'script-rule-matched': <T>(rule: CosmeticFilter, context: EngineEventContext<T>) => any;
+  'extended-rule-matched': <T>(rule: CosmeticFilter, context: EngineEventContext<T>) => any;
+  'style-rule-matched': <T>(rule: CosmeticFilter, context: EngineEventContext<T>) => any;
```

Comment:
What is the returned value expected to be?

Reply:
Nothing, or everything. Because people often add a branch returning a meaningless value like:

```js
function callback() {
  if (false) {
    return anotherFunctionReturningValue();
  }

  ...
}
```